### ISSUE TITLE
CP01 and CP04 can cause clashing treatment of NULL within a single file

### ISSUE THREAD

**Issue by PaddyAlton** [sqlfluff/sqlfluff#5099]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I have set up `sqlfluff` to render keywords as uppercase and literals as lowercase. It seems that in certain places, `NULL` is treated as a keyword, and in others `null` is treated as a literal.

The result: I have files where the linter wants lowercase `null` for e.g. `CAST(null AS string)` and uppercase `NULL` for e.g. `some_field IS NULL`.

### Expected Behaviour

I would expect my ruleset to make all instances of `null` lowercase. I have the following set in `.sqlfluff`:

```
[sqlfluff:rules:capitalisation.literals]
capitalisation_policy = lower
```

Personally I think `null` should be treated as a literal. [The docs indicate that it should be](https://docs.sqlfluff.com/en/stable/rules.html#rule-capitalisation.literals). I would expect consistent capitalisation to be applied to my SQL files.

(I appreciate the nuance that e.g. in `IGNORE NULLS`, `NULLS` would be a keyword)

### Observed Behaviour

If I try to force consistent capitalisation, I get the following linting errors:

### `field_a IS null`
```
L:   4 | P:  16 | CP01 | Keywords must be upper case.
                       | [capitalisation.keywords]
```

### `CAST(NULL AS string)`
```
L:   5 | P:  10 | CP04 | Boolean/null literals must be lower case.
                       | [capitalisation.literals]
```

### How to reproduce

Here is a minimal SQL file, shown after running `sqlfluff fix test.sql` to change it to a format accepted by the linter. Note the inconsistent capitalisation:

```sql
-- test.sql

SELECT
    field_a IS NULL AS nullcheck,
    CAST(null AS string) AS unknown_string,
FROM
    some_table
```

To reproduce, change the capitalisation of either (or both) 'null's and run `sqlfluff lint test.sql`.

### Dialect

BigQuery

### Version

sqlfluff 2.3.0
python 3.11.4
dbt 1.6.0
sqlfluff-templater-dbt 2.3.0

### Configuration

```
[sqlfluff]
templater = dbt
dialect = bigquery
exclude_rules = LT08

[sqlfluff:rules:aliasing.table]
aliasing.table = explicit

[sqlfluff:rules:aliasing.column]
aliasing.column = explicit

[sqlfluff:rules:capitalisation.keywords]
capitalisation_policy = upper
[sqlfluff:rules:capitalisation.identifiers]
capitalisation_policy = lower
[sqlfluff:rules:capitalisation.functions]
extended_capitalisation_policy = upper
[sqlfluff:rules:capitalisation.literals]
capitalisation_policy = lower
[sqlfluff:rules:capitalisation.types]
extended_capitalisation_policy = lower

[sqlfluff:rules:convention.select_trailing_comma]
select_clause_trailing_comma = require
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by alanmcruickshank** [comment]
Agreed. Good catch, for some reason the `null` is being parsed as a keyword in the `IS` expression there. I've confirmed the issue with this query:

```sql
SELECT
    CAST(null AS string),
    field IS null AS foo
FROM my_table
```

And running `sqlfluff parse` I get this result:

```
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        select_statement:
[L:  1, P:  1]      |            select_clause:
[L:  1, P:  1]      |                keyword:                                      'SELECT'
[L:  1, P:  7]      |                [META] indent:
[L:  1, P:  7]      |                newline:                                      '\n'
[L:  2, P:  1]      |                whitespace:                                   '    '
[L:  2, P:  5]      |                select_clause_element:
[L:  2, P:  5]      |                    function:
[L:  2, P:  5]      |                        function_name:
[L:  2, P:  5]      |                            function_name_identifier:         'CAST'
[L:  2, P:  9]      |                        bracketed:
[L:  2, P:  9]      |                            start_bracket:                    '('
[L:  2, P: 10]      |                            [META] indent:
[L:  2, P: 10]      |                            expression:
[L:  2, P: 10]      |                                null_literal:                 'null'
[L:  2, P: 14]      |                            whitespace:                       ' '
[L:  2, P: 15]      |                            keyword:                          'AS'
[L:  2, P: 17]      |                            whitespace:                       ' '
[L:  2, P: 18]      |                            data_type:
[L:  2, P: 18]      |                                data_type_identifier:         'string'
[L:  2, P: 24]      |                            [META] dedent:
[L:  2, P: 24]      |                            end_bracket:                      ')'
[L:  2, P: 25]      |                comma:                                        ','
[L:  2, P: 26]      |                newline:                                      '\n'
[L:  3, P:  1]      |                whitespace:                                   '    '
[L:  3, P:  5]      |                select_clause_element:
[L:  3, P:  5]      |                    expression:
[L:  3, P:  5]      |                        column_reference:
[L:  3, P:  5]      |                            naked_identifier:                 'field'
[L:  3, P: 10]      |                        whitespace:                           ' '
[L:  3, P: 11]      |                        keyword:                              'IS'
[L:  3, P: 13]      |                        whitespace:                           ' '
[L:  3, P: 14]      |                        keyword:                              'null'
[L:  3, P: 18]      |                    whitespace:                               ' '
[L:  3, P: 19]      |                    alias_expression:
[L:  3, P: 19]      |                        [META] indent:
[L:  3, P: 19]      |                        keyword:                              'AS'
[L:  3, P: 21]      |                        whitespace:                           ' '
[L:  3, P: 22]      |                        naked_identifier:                     'foo'
[L:  3, P: 25]      |                        [META] dedent:
[L:  3, P: 25]      |            newline:                                          '\n'
[L:  4, P:  1]      |            [META] dedent:
[L:  4, P:  1]      |            from_clause:
[L:  4, P:  1]      |                keyword:                                      'FROM'
[L:  4, P:  5]      |                whitespace:                                   ' '
[L:  4, P:  6]      |                from_expression:
[L:  4, P:  6]      |                    [META] indent:
[L:  4, P:  6]      |                    from_expression_element:
[L:  4, P:  6]      |                        table_expression:
[L:  4, P:  6]      |                            table_reference:
[L:  4, P:  6]      |                                naked_identifier:             'my_table'
[L:  4, P: 14]      |                    [META] dedent:
[L:  4, P: 14]      |    newline:                                                  '\n'
[L:  5, P:  1]      |    [META] end_of_file:
```

It shows the first `null` being parsed as a null literal, and the second being parsed as a keyword. I'm hoping this should be an easy fix.
